### PR TITLE
DNS Proxy: Allow SO_LINGER to be set to the socket to upstream

### PR DESCRIPTION
**Pull request by gandro**
Add a flag to allow setting the `SO_LINGER` socketopt on the socket for the connection of DNS proxy to the upstream DNS server. The flag is a tristate: -1 to disable, 0 to send TCP RST on close, and a timeout value (in seconds) to let close block at most.

Setting linger can positively impact the operation of Cilium's DNS proxy due to the following circumstances:

1. When a DNS response is large, the server can truncate its reply. Using the truncated flag on DNS responses, the server can signal to the client that it did not receive the full response, and that it should retry the query using TCP.

2. DNS clients often simultaneously request both A and AAAA records. Typically, they do so from the same source port. When using TCP, both queries are sent via the same, persistent TCP connection.

3. Cilium's DNS proxy fails to correctly handle this persistent TCP connection - it opens a connection to the upstream DNS server _per query_.

4. When running in transparent mode, the DNS proxy thus may attempt to bind the same local source IP/port combination rapidly - this only works since the proxy sets the SO_REUSEADDR socket opt. Unfortunately, depending on network circumstances, the second bind-then-connect can fail [due to a kernel bug](https://github.com/cilium/cilium/issues/31197#issuecomment-2209277800)

Setting the linger timeout to zero has the effect of sending a TCP RST as soon as the connection is closed - i.e. forcefully slamming the connection shut. This works around the issue of DNS proxy's errors when reusing the same local ip/port combination in a relatively non-invasive way, since it avoid the kernel bug around socket reuse. However, it doesn't solve the root cause of the proxy attempting to open two connections when it should persist the connection just like the client does.

Co-developed with @bimmlerd

### CODE REVIEW

**Review by pippolo84 — APPROVED** [review]
Great, thanks! :100:

File: pkg/defaults/defaults.go [hunk 1]
```diff
@@ -189,6 +189,10 @@ const (
 	// DNSProxyLockCount.
 	DNSProxyLockTimeout = 500 * time.Millisecond
 
+	// DNSProxySocketLingerTimeout defines how long we wait for the connection
+	// between the DNS proxy and the upstream server to be closed.
+	DNSProxySocketLingerTimeout = 10
```
Comment (derailed) [inline pkg/defaults/defaults.go:194]: nit: Might be good here to specify the unit (second) in the comment for posterity.

**Review by derailed — APPROVED** [review]
@gandro  @bimmlerd  Nice work!

**Review by youngnick — APPROVED** [review]
LGTM from sig-k8s.

**Review by squeed — APPROVED** [review]

**Pull request merged**